staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go

```diff
@@ -203,9 +205,11 @@ type udsGRPCConnector struct {
 // TODO At the moment, it does not use the provided context.
 func (u *udsGRPCConnector) connect(_ context.Context) (proxier, error) {
 	udsName := u.udsName
-	dialOption := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+	dialOption := grpc.WithContextDialer(func(grpcDialCtx context.Context, addr string) (net.Conn, error) {
```

Comment:
Alternative proposal worth considering:

Add `grpc.WithBlock()` and `grpc.WithTimeout()` as dial options passed into `CreateSingleUseGrpcTunnel()`. This way the timeout between the request context and the dial is more distinct

Reply:
This one might be more tricky since we would have to consider it alongside kube-apiserver retry logic for webhook requests, but setting `FailOnNonTempDialError(true)` as a dial option might be worth considering as well. This way the actual error when dialing is surfaced instead of a timeout error or a "context deadline exceeded" error.

Reply:
I ended up going a different way - modify at "proxyServerConnector" level. In the UDS+grpc case this will be the primary ctx passed to grpc.DialContext, and propagate to underlying (such as UDS dialer).

> Add grpc.WithBlock()

I added this. I'm not sure it changes much in terms of reliability. It should make the dial_failure metric for StageConnect more accurate.

> FailOnNonTempDialError(true)

This one is marked experimental, which makes me hesitate.

Reply:
> This one is marked experimental, which makes me hesitate.

Yeah this is unfortunate. It would be nice to to surface these errors somehow, because even with the timeout and the added metric from https://github.com/kubernetes/kubernetes/pull/113520, we still wouldn't have any visibility for why this dial is timing out and what errors are being encountered on each retry.

Reply:
I'm convinced, and added this in latest commit push. It is available very early (grpc v1.0.1) so won't prevent any backport, and this is pretty low risk (blast radius is a single egress request).